### PR TITLE
perf (`TreeMap.range`): Update the TreeMap->Range logic

### LINKED ISSUE #928: Optimize `TreeMap` range query implementation
`TreeMap` is awkward to use as its range query complexity is `O(Klog(N))`, which means iterating it is very costly. This happens because `tree_map::iter::Keys` holds a bounded range defining the iterator elements, but not their position in the tree:

```rust
pub struct Keys<'a, K: 'a>
where
    K: BorshSerialize + BorshDeserialize + Ord,
{
    tree: &'a Tree<K>,
    length: u32,
    min: Bound<&'a K>,
    max: Bound<&'a K>,
}
```

So every time we search for the next (or previous) element in the tree we have to search it from the top of the tree:
```rust
fn next(&mut self) -> Option<&'a K> {
    ...
    let next = next_asc(self.tree, self.min);
    ...
}

fn next_asc<'a, K>(tree: &'a Tree<K>, bound: Bound<&'a K>) -> Option<&'a K>
where
    K: BorshSerialize + Ord + BorshDeserialize,
{
    match bound {
        Bound::Unbounded => tree.min(),
        Bound::Included(bound) => tree.ceil_key(bound),
        Bound::Excluded(bound) => tree.higher(bound),
    }
}
```

This does not have to be the case, the range query can be answered in `O(K + log(N))` if we were to save the actual bounding tree nodes and find the next (or previous) element in `O(1)` (amortized). This cannot be easily done with the current implementation however as there is no way to traverse up to the parent for a given node. So a decent refactoring is required, but we could actually avoid storing the parent index in the `Node` struct if we hold a `cur_path: Vec<FreeListIndex>` in the iterator instead (i.e. a list of node indices we have gone through to reach the current node).

Part of #818

### PR DESCRIPTION
**Pull request by askoa**
Closes #928 

Update the `TreeMap->Range` logic to keep track of current path. See the Issue description for detailed explanation.

### CODE REVIEW

File: near-sdk/src/store/tree_map/iter.rs
Comment (itegulov) [inline near-sdk/src/store/tree_map/iter.rs:604]:
```suggestion
            if let Some(result_idx) = self.stack_asc.last() {
```

File: near-sdk/src/store/tree_map/iter.rs
Comment (itegulov) [inline near-sdk/src/store/tree_map/iter.rs:647]:
```suggestion
            if let Some(result_idx) = self.stack_desc.last() {
```

**Review by itegulov — COMMENTED** [review]
The implementation looks solid, great job! The same should be done for `Keys` (in the same file) and we should be good.

**Comment by askoa** [comment]
>  The same should be done for `Keys` (in the same file) and we should be good.

Done.

**Comment by askoa** [comment]
The build failure seems unrelated to this PR.

```
thread 'actix-rt|system:0|arbiter:1' panicked at 'failed to start listening on server_addr=0.0.0.0:19718 e=Os { code: 98, kind: AddrInUse, message: "Address already in use" }', chain/network/src/peer_manager/peer_manager_actor.rs:302:29
```

**Review by itegulov — APPROVED** [review]
Ran some benchmarks and it does seem like this implementation's asymptotic is better than the old one. Source: https://github.com/itegulov/storage_collections_bench/tree/tree-map-iter. The farthest I could test without hitting the max gas limit was ~32 insert operations + ~32 iterations, where this implementation is 20% more efficient.

@austinabell you have the most context on this, so could you take a look before we merge this just in case?

**Comment by itegulov** [comment]
> The build failure seems unrelated to this PR.

Yeah, don't worry about it. Just some flaky workspaces behaviour.

File: near-sdk/src/store/tree_map/iter.rs
Comment (austinabell) [inline near-sdk/src/store/tree_map/iter.rs:731]: At a glance, this logic seems identical to that for `Keys`. What is preventing it from being abstracted to be shared between the two?

**Review by austinabell — APPROVED** [review]
Everything should be good to go. I think existing tests are exhaustive enough and if the benchmarks are showing improvement, that's probably good enough!

The only comment is just on the duplication of the logic, but that refactor could happen in a follow-up PR if needed

**Comment by austinabell** [comment]
> where this implementation is 20% more efficient.

Honestly would have expected this to be more of an optimization. I suppose because now since the values are cached anyway, it's only extra computing that is the overhead whereas if this was done for `sdk::collections::TreeMap` the performance gain would be much more? (Asking out of curiosity)

File: near-sdk/src/store/tree_map/iter.rs
Comment (askoa) [reply near-sdk/src/store/tree_map/iter.rs:731]: `Keys` seems to be special case of `KeysRange` and can be easily replaced. May be a followup Issue to remove `Keys` and replace with `KeysRange` might be a better solution. What do you think?

Comment (austinabell) [reply near-sdk/src/store/tree_map/iter.rs:731]: yeah, that's fine, can do in a follow up!

@itegulov feel free to merge this if you're happy with it, unsure if you had any other thoughts since you seem to be the stakeholder for that issue

Comment (askoa) [reply near-sdk/src/store/tree_map/iter.rs:731]: Created #987 to track the issue.

**Comment by itegulov** [comment]
@askoa thanks again for handling this!

**Pull request merged**
